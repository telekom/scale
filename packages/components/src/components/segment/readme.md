# scale-segmented-button



<!-- Auto Generated Below -->


## Properties

| Property                     | Attribute                      | Description                                                                                                                                            | Type                               | Default     |
| ---------------------------- | ------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------ | ---------------------------------- | ----------- |
| `adjacentSiblings`           | `adjacent-siblings`            |                                                                                                                                                        | `"left" \| "leftright" \| "right"` | `undefined` |
| `ariaDescriptionTranslation` | `aria-description-translation` | a11y text for getting meaningful value. `$buttonNumber` and `$selected` are template variables and will be replaces by their corresponding properties. | `string`                           | `''`        |
| `disabled`                   | `disabled`                     | (optional) If `true`, the button is disabled                                                                                                           | `boolean`                          | `false`     |
| `iconAriaLabel`              | `icon-aria-label`              | (optional) Icon aria-label for icon only                                                                                                               | `string`                           | `undefined` |
| `iconOnly`                   | `icon-only`                    | (optional) icon only segment                                                                                                                           | `boolean`                          | `false`     |
| `iconText`                   | `icon-text`                    | (optional) segment with icon and text                                                                                                                  | `boolean`                          | `false`     |
| `multiSelect`                | `multi-select`                 | (optional) multi select segment                                                                                                                        | `boolean`                          | `false`     |
| `position`                   | `position`                     | (optional) position within group                                                                                                                       | `number`                           | `undefined` |
| `segmentId`                  | `segment-id`                   | (optional) segment's id                                                                                                                                | `string`                           | `undefined` |
| `selected`                   | `selected`                     | (optional) If `true`, the button is selected                                                                                                           | `boolean`                          | `false`     |
| `size`                       | `size`                         | (optional) The size of the button                                                                                                                      | `"large" \| "medium" \| "small"`   | `'small'`   |
| `styles`                     | `styles`                       | (optional) Injected CSS styles                                                                                                                         | `string`                           | `undefined` |
| `width`                      | `width`                        | (optional) Button width set to ensure that all buttons have the same width                                                                             | `string`                           | `undefined` |


## Events

| Event         | Description                                                                                        | Type                                              |
| ------------- | -------------------------------------------------------------------------------------------------- | ------------------------------------------------- |
| `scale-click` | Emitted when button is clicked                                                                     | `CustomEvent<{ id: string; selected: boolean; }>` |
| `scaleClick`  | <span style="color:red">**[DEPRECATED]**</span> in v3 in favor of kebab-case event names<br/><br/> | `CustomEvent<{ id: string; selected: boolean; }>` |


## Methods

### `setFocus() => Promise<void>`



#### Returns

Type: `Promise<void>`




## Dependencies

### Depends on

- [scale-icon-action-success](../icons/action-success)

### Graph
```mermaid
graph TD;
  scale-segment --> scale-icon-action-success
  style scale-segment fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
