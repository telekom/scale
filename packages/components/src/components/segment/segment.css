/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  /* not selected */
  --background-color-selected: var(--telekom-color-ui-base);
  --button-radius: 6px;
  --padding-top-bottom: var(--telekom-spacing-composition-space-03);
  --height: 28px;
  --font: var(--telekom-text-style-caption-bold);
  --label-disabled: var(--telekom-color-text-and-icon-disabled);
  --transition: all var(--telekom-motion-duration-transition)
    var(--telekom-motion-easing-standard);
  --color: var(--telekom-color-text-and-icon-standard);

  --color-selected: var(--telekom-color-text-and-icon-primary-standard);

  /* medium styles */
  --font-medium: var(--telekom-text-style-ui-bold);
  --height-medium: 36px;

  /* large styles */
  --font-large: var(--telekom-text-style-ui-bold);
  --height-large: 40px;

  /* hover styles */
  --background-color-hover: var(--telekom-color-ui-state-fill-hovered);

  /* active styles */
  --background-color-active: var(--telekom-color-ui-state-fill-pressed);

  display: flex;
  align-items: center;
  justify-content: center;
  flex: 1;
}

.segment {
  box-sizing: border-box;
  position: relative;
  border: 0;
  height: 100%;
  width: 100%;
  border-radius: var(--button-radius);
  padding-top: var(--padding-top-bottom);
  padding-bottom: var(--padding-top-bottom);
  transition: var(--transition);
  background-color: transparent;
  padding: 0;
}

.segment--mask {
  box-sizing: border-box;
  display: flex;
  align-items: center;
  justify-content: center;
  position: relative;
  height: 100%;
  width: 100%;
  border-radius: var(--button-radius);
  background-color: transparent;
  font: var(--font);
  transition: var(--transition);
  white-space: nowrap;
  color: var(--color);
  line-height: 1;
}

button.segment--small.segment--selected .segment--mask {
  padding-left: var(--telekom-spacing-composition-space-04);
  padding-right: var(--telekom-spacing-composition-space-05);
}

button.segment--icon-only.segment--small .segment--mask {
  padding-left: var(--telekom-spacing-composition-space-06);
  padding-right: var(--telekom-spacing-composition-space-06);
}

button.segment--small .segment--mask {
  padding-left: 0.625rem;
  padding-right: 0.625rem;
}

.segment--small {
  height: var(--height);
}

.segment--medium {
  height: var(--height-medium);
}

button.segment--medium.segment--selected .segment--mask {
  padding-left: var(--telekom-spacing-composition-space-05);
  padding-right: var(--telekom-spacing-composition-space-06);
}

button.segment--medium .segment--mask {
  font: var(--font-medium);
  padding-left: 0.875rem;
  padding-right: 0.875rem;
}

.segment--large {
  height: var(--height-large);
}

button.segment--large.segment--selected .segment--mask {
  padding-left: var(--telekom-spacing-composition-space-06);
  padding-right: var(--telekom-spacing-composition-space-07);
}

button.segment--large .segment--mask {
  font: var(--font-large);
  padding-left: var(--telekom-spacing-composition-space-06);
  padding-right: var(--telekom-spacing-composition-space-07);
}

.segment--selected {
  background-color: var(--background-color-selected);
  box-shadow: var(--telekom-shadow-raised-standard);
}

.segment--selected .segment--mask {
  color: var(--color-selected);
}

.segment--disabled.segment--selected .segment--mask {
  color: var(--label-disabled);
}

.segment--disabled:not(.segment--selected) .segment--mask {
  color: var(--label-disabled);
  background-color: transparent;
}

.segment--left-sibling-selected {
  border-radius: 0 var(--button-radius) var(--button-radius) 0;
  margin-left: 0;
  clip-path: inset(-2.75rem -2.75rem -2.75rem 0);
  backface-visibility: hidden;
  overflow: hidden;
}

.segment--right-sibling-selected {
  border-radius: var(--button-radius) 0 0 var(--button-radius);
  clip-path: inset(-2.75rem 0 -2.75rem -2.75rem);
  margin-right: 0;
  backface-visibility: hidden;
  overflow: hidden;
  position: relative;
}

.segment--left-right-sibling-selected {
  clip-path: inset(-2.75rem 0 -2.75rem 0);
  border-radius: 0;
  margin-left: 0;
  margin-right: 0;
  backface-visibility: hidden;
}

.segment:not(.segment--disabled):hover .segment--mask {
  background-color: var(--background-color-hover);
}

.segment:not(.segment--disabled):active .segment--mask {
  background-color: var(--background-color-active);
}

.segment:focus {
  box-shadow: var(--telekom-shadow-raised-standard);
  outline: var(--telekom-spacing-composition-space-02) solid
    var(--telekom-color-functional-focus-standard);
  outline-offset: var(--telekom-spacing-composition-space-01);
  z-index: 40;
  position: relative;
}

.segment:not(.segment--selected):focus {
  outline-offset: 0;
}

.segment--left-right-sibling-selected:focus,
.segment--right-sibling-selected:focus,
.segment--left-sibling-selected:focus {
  clip-path: none;
}

/* FIXME remove the hard-coded icon names */
.segment:not(.segment--icon-only) scale-icon-action-checkmark,
.segment scale-icon-action-close {
  margin-right: var(--telekom-spacing-composition-space-03);
}

.segment:not(.segment--selected) scale-icon-action-checkmark {
  visibility: hidden !important;
}

.segment scale-icon-action-checkmark {
  vertical-align: middle;
}

.segment--selected:not(.segment--icon-only) .icon-container {
  display: none !important;
}

.segment .success-icon-container {
  display: none;
}

.segment--selected .success-icon-container {
  display: flex;
  justify-content: center;
  align-items: center;
}

.icon-container {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.segment--selected .icon-container {
  color: var(--color-selected);
}
