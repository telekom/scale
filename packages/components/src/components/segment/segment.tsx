/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import {
  Component,
  Prop,
  h,
  Host,
  Element,
  Event,
  EventEmitter,
  Method,
} from '@stencil/core';
import classNames from 'classnames';
import { emitEvent } from '../../utils/utils';
import statusNote from '../../utils/status-note';

let i = 0;

@Component({
  tag: 'scale-segment',
  styleUrl: 'segment.css',
  shadow: true,
})
export class Segment {
  @Element() hostElement: HTMLElement;
  /** (optional) The size of the button */
  @Prop() size?: 'small' | 'medium' | 'large' = 'small';
  /** (optional) If `true`, the button is selected */
  @Prop({ mutable: true }) selected?: boolean = false;
  /** (optional) If `true`, the button is disabled */
  @Prop() disabled?: boolean = false;
  /** (optional) segment's id */
  @Prop({ reflect: true, mutable: true }) segmentId?: string;
  /** (optional) aria-label attribute needed for icon-only buttons */
  @Prop() ariaLabelSegment: string;
  /** (optional) Button width set to ensure that all buttons have the same width */
  @Prop() width?: string;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;
  // /** (optional)  */
  @Prop({ reflect: true, mutable: true }) adjacentSiblings?:
    | 'left'
    | 'right'
    | 'leftright';
  /** (optional) translation of 'selected */
  @Prop() ariaLangSelected? = 'selected';
  /** (optional) translation of 'deselected */
  @Prop() ariaLangDeselected? = 'deselected';
  /** a11y text for getting meaningful value. `$buttonNumber` and `$selected` are template variables and will be replaces by their corresponding properties.  */
  @Prop() ariaDescriptionTranslation = '$selected';
  /** (optional) position within group */
  @Prop() position?: number;
  /** (optional) position within group */
  @Prop({ mutable: true }) hasIcon?: boolean;
  /** (optional) position within group */
  @Prop({ mutable: true }) textOnly?: boolean;
  /** (optional) position within group */
  @Prop({ mutable: true }) iconOnly?: boolean;
  /** Emitted when button is clicked */
  @Event({ eventName: 'scale-click' }) scaleClick!: EventEmitter<{
    id: string;
    selected: boolean;
  }>;
  /** @deprecated in v3 in favor of kebab-case event names */
  @Event({ eventName: 'scaleClick' }) scaleClickLegacy!: EventEmitter<{
    id: string;
    selected: boolean;
  }>;

  private focusableElement: HTMLElement;

  @Method()
  async setFocus() {
    this.focusableElement.focus();
  }

  componentDidRender() {
    if (this.hostElement.hasAttribute('aria-label')) {
      statusNote({
        tag: 'deprecated',
        message:
          'Property "ariaLabel" is deprecated. Please use the "ariaLabelSegment" property!',
        type: 'warn',
        source: this.hostElement,
      });
    }
  }

  componentWillLoad() {
    if (this.segmentId == null) {
      this.segmentId = 'segment-' + i++;
    }
  }
  componentDidUpdate() {
    this.handleIcon();
  }

  getAriaDescriptionTranslation() {
    const replaceSelected = this.selected
      ? this.ariaLangSelected
      : this.ariaLangDeselected;
    const filledText = this.ariaDescriptionTranslation
      .replace(/\$position/g, `${this.position}`)
      .replace(/\$selected/g, `${replaceSelected}`);
    return filledText;
  }

  handleIcon() {
    Array.from(this.hostElement.childNodes).forEach((child) => {
      if (
        child.nodeType == 1 &&
        child.nodeName.substr(0, 10) === 'SCALE-ICON'
      ) {
        const icon: HTMLElement = this.hostElement.querySelector(
          child.nodeName
        );
        switch (this.size) {
          case 'small':
            icon.setAttribute('size', '14');
            break;
          case 'medium' || 'large':
            icon.setAttribute('size', '16');
            break;
        }
        icon.style.display = 'inline-flex';
        icon.style.marginRight = '4px';
        this.hasIcon = true;
      }
      if (child.nodeType == 3 && this.hostElement.childNodes.length == 1) {
        this.textOnly = true;
        var span = document.createElement('span');
        child.parentNode.insertBefore(span, child);
        span.appendChild(child);
      }
      if (
        child.nodeType == 1 &&
        child.nodeName.substr(0, 10) === 'SCALE-ICON' &&
        this.hostElement.childNodes.length === 1
      ) {
        this.iconOnly = true;
        this.hostElement.setAttribute('icon-only', 'true');
        const icon: HTMLElement = this.hostElement.querySelector(
          child.nodeName
        );
        icon.style.marginRight = '0px';
        this.selected
          ? icon.setAttribute('selected', '')
          : icon.removeAttribute('selected');
      }
    });
  }

  handleClick = (event: MouseEvent) => {
    event.preventDefault();
    this.selected = !this.selected;
    emitEvent(this, 'scaleClick', {
      id: this.segmentId,
      selected: this.selected,
    });
  };

  render() {
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <button
          ref={(el) => (this.focusableElement = el)}
          class={this.getCssClassMap()}
          id={this.segmentId}
          onClick={this.handleClick}
          disabled={this.disabled}
          type="button"
          style={{ width: this.width }}
          aria-label={this.ariaLabelSegment}
          aria-pressed={this.selected}
          part={this.getBasePartMap()}
          aria-description={this.getAriaDescriptionTranslation()}
        >
          <div class="segment--mask">
            {!this.iconOnly && (
              <div class="success-icon-container">
                <scale-icon-action-success
                  size={this.size === 'small' ? 14 : 16}
                  class="scale-icon-action-success"
                  accessibility-title="success"
                  selected
                />
              </div>
            )}
            <div class="icon-container">
              <slot name="segment-icon" />
            </div>
            <slot />
          </div>
        </button>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const prefix = mode === 'basePart' ? '' : 'segment--';

    return classNames(
      'segment',
      this.size && `${prefix}${this.size}`,
      this.selected && `${prefix}selected`,
      this.disabled && `${prefix}disabled`,
      this.adjacentSiblings &&
        `${prefix}${this.adjacentSiblings.replace(/ /g, '-')}-sibling-selected`,
      this.hasIcon && `${prefix}has-icon`,
      this.iconOnly && `${prefix}icon-only`
    );
  }
}
