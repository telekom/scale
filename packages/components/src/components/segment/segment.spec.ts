import { newSpecPage } from '@stencil/core/testing';
import { Segment } from './segment';

describe('Segment', () => {
  it('should match standard snapshot', async () => {
    const page = await newSpecPage({
      components: [Segment],
      html: `
        <scale-segment>Label</scale-segment>`,
    });
    expect(page.root).toMatchSnapshot();
  });
  it('should include right properties', async () => {
    const page = await newSpecPage({
      components: [Segment],
      html: `
            <scale-segment>Label</scale-segment>`,
    });
    page.root.disabled = true;
    (page.root.size = 'medium'), (page.root.selected = true);
    page.root.hasIcon = true;
    page.root.ariaLangSelected = 'ariaLangSelected';
    page.root.ariaLangDeselected = 'ariaLangDeselected';
    page.root.ariaDescriptionTranslation = 'ariaDescriptionTranslation';
    await page.waitForChanges();
    expect(page.rootInstance.disabled).toBe(true);
    expect(page.rootInstance.size).toBe('medium');
    expect(page.rootInstance.selected).toBe(true);
    expect(page.rootInstance.hasIcon).toBe(true);
    expect(page.rootInstance.ariaLangSelected).toBe('ariaLangSelected');
    expect(page.rootInstance.ariaLangDeselected).toBe('ariaLangDeselected');
    expect(page.rootInstance.ariaDescriptionTranslation).toBe(
      'ariaDescriptionTranslation'
    );
  });
  it('should include iconOnly', async () => {
    const page = await newSpecPage({
      components: [Segment],
      html: `
            <scale-segment>Label</scale-segment>`,
    });
    page.root.iconOnly = true;
    await page.waitForChanges();
    expect(page.rootInstance.iconOnly).toBe(true);
  });
  it('should include textOnly', async () => {
    const page = await newSpecPage({
      components: [Segment],
      html: `
            <scale-segment>Label</scale-segment>`,
    });
    page.root.textOnly = true;
    await page.waitForChanges();
    expect(page.rootInstance.textOnly).toBe(true);
  });
});
