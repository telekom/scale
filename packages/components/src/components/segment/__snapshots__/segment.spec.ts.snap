// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Segment should match standard snapshot 1`] = `
<scale-segment segment-id="segment-0">
  <mock:shadow-root>
    <button aria-description="deselected" class="segment segment--small" id="segment-0" part="segment small" type="button">
      <div class="segment--mask">
        <div class="success-icon-container">
          <scale-icon-action-success accessibility-title="success" class="scale-icon-action-success" selected="" size="14"></scale-icon-action-success>
        </div>
        <div class="icon-container">
          <slot name="segment-icon"></slot>
        </div>
        <slot></slot>
      </div>
    </button>
  </mock:shadow-root>
  Label
</scale-segment>
`;
